src/V3Trace.cpp

```diff
@@ -372,79 +372,99 @@ class TraceVisitor : public EmitCBaseVisitor {
-        AstCFunc* funcp = NULL;
-        if (basep->funcType() == AstCFuncType::TRACE_FULL) {
-            funcp = newCFunc(AstCFuncType::TRACE_FULL_SUB, name, basep);
-        } else if (basep->funcType() == AstCFuncType::TRACE_CHANGE) {
-            funcp = newCFunc(AstCFuncType::TRACE_CHANGE_SUB, name, basep);
-        } else {
-            basep->v3fatalSrc("Strange base function type");
+        funcp->declPrivate(true);
+        // Add it to top scope
+        m_topScopep->addActivep(funcp);
+        // Add call to new function
+        if (callfromp) {
+            AstCCall* callp = new AstCCall(funcp->fileline(), funcp);
+            callp->argTypes("userp, tracep");
+            callfromp->addStmtsp(callp);
         }
-        AstCCall* callp = new AstCCall(funcp->fileline(), funcp);
-        callp->argTypes("vlSymsp, vcdp, code");
-        if (VN_IS(callfromp, CFunc)) {
-            VN_CAST(callfromp, CFunc)->addStmtsp(callp);
-        } else if (VN_IS(callfromp, If)) {
-            VN_CAST(callfromp, If)->addIfsp(callp);
-        } else {
-            callfromp->v3fatalSrc("Unknown caller node type");  // Where to add it??
+        // Register function
+        if (regp) {
+            string registration = "tracep->add";
+            if (type == AstCFuncType::TRACE_FULL) {
+                registration += "Full";
+            } else if (type == AstCFuncType::TRACE_CHANGE) {
+                registration += "Chg";
+            } else {
+                funcp->v3fatal("Don't know how to register this type of function");
+            }
+            registration += "Cb(&" + protect(name) + ", __VlSymsp);\n";
+            AstCStmt* const stmtp = new AstCStmt(flp, registration);
+            regp->addStmtsp(stmtp);
         }
+        // Add global activity check to TRACE_CHANGE functions
+        if (type == AstCFuncType::TRACE_CHANGE) {
+            funcp->addInitsp(
+                new AstCStmt(flp, string("if (VL_UNLIKELY(!vlSymsp->__Vm_activity)) return;\n")));
+        }
+        // Done
+        UINFO(5, "  newCFunc " << funcp << endl);
         return funcp;
     }
-    void addToChgSub(AstNode* underp, AstNode* stmtsp) {
-        if (!m_chgSubFuncp || (m_chgSubParentp != underp)
-            || (m_chgSubStmts && v3Global.opt.outputSplitCTrace()
-                && m_chgSubStmts > v3Global.opt.outputSplitCTrace())) {
-            m_chgSubFuncp = newCFuncSub(m_chgFuncp, underp);
-            m_chgSubParentp = underp;
-            m_chgSubStmts = 0;
+
+    void assignDeclCode(AstTraceDecl* nodep) {
+        if (!nodep->code()) {
+            nodep->code(m_code);
+            m_code += nodep->codeInc();
+            m_statUniqCodes += nodep->codeInc();
+            ++m_statUniqSigs;
         }
-        m_chgSubFuncp->addStmtsp(stmtsp);
-        m_chgSubStmts += EmitCBaseCounterVisitor(stmtsp).count();
     }
 
-    void putTracesIntoTree() {
+    void createTraceFunctions() {
```

Comment:
This function appears crazy-long, please split it up, ideally up to 50, 100ish lines at most in a function so it's easier to grok.  Your //// might be good split points.